rio/src/RIO/File.hs

```diff
@@ -0,0 +1,345 @@
+import           System.FilePath        (takeDirectory, takeFileName, (</>))
+import           System.Posix.Internals (CFilePath, c_close, c_safe_open,
+                                         withFilePath)
+import           System.Posix.Types     (CMode (..), Fd (..))
+
+#if MIN_VERSION_base(4,9,0)
+import qualified GHC.IO.Handle.Types    as HandleFD (Handle (..), Handle__ (..))
+#endif
+
+
+-- TODO: Add a ticket/pull request to export this symbols from
+-- System.Internal.Posix
+--
+-- NOTE: System.Posix.Internal doesn't re-export this constants so we have to
+-- recreate-them here
+foreign import ccall unsafe "HsBase.h __hscore_o_rdonly" o_RDONLY :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_wronly" o_WRONLY :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_rdwr"   o_RDWR   :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_append" o_APPEND :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_creat"  o_CREAT  :: CInt
+foreign import ccall unsafe "HsBase.h __hscore_o_noctty" o_NOCTTY :: CInt
+
+-- After here, we have our own imports
+foreign import ccall safe "fcntl.h openat"
+  c_safe_openat :: CInt -> CFilePath -> CInt -> CMode -> IO CInt
+
+foreign import ccall safe "fcntl.h renameat"
+  c_safe_renameat :: CInt -> CFilePath -> CInt -> CFilePath -> IO CInt
+
+foreign import ccall safe "unistd.h fsync"
+  c_safe_fsync :: CInt -> IO CInt
+
+std_flags, output_flags, read_flags, write_flags, rw_flags,
+    append_flags :: CInt
+std_flags    = o_NOCTTY
+output_flags = std_flags    .|. o_CREAT
+read_flags   = std_flags    .|. o_RDONLY
+write_flags  = output_flags .|. o_WRONLY
+rw_flags     = output_flags .|. o_RDWR
+append_flags = write_flags  .|. o_APPEND
+
+ioModeToFlags :: IOMode -> CInt
+ioModeToFlags iomode =
+  case iomode of
+    ReadMode      -> read_flags
+    WriteMode     -> write_flags
+    ReadWriteMode -> rw_flags
+    AppendMode    -> append_flags
+
+openDir :: MonadIO m => FilePath -> m Fd
+openDir fp
+  -- TODO: Investigate what is the situation with Windows FS in regards to non_blocking
+  -- NOTE: File operations _do not support_ non_blocking on various kernels, more
+  -- info can be found here: https://ghc.haskell.org/trac/ghc/ticket/15153
+ =
+  liftIO $
+  withFilePath fp $ \cFp ->
+    Fd <$>
+    (throwErrnoIfMinus1Retry "openDir" $
+     c_safe_open cFp (ioModeToFlags ReadMode) 0o660)
+
+-- NOTE: Not Async exception safe
+openFileFromDir :: (MonadIO m) => Fd -> FilePath -> IOMode -> m Handle
+openFileFromDir (Fd dirFd) fp iomode =
+  liftIO $
+  withFilePath fp $ \f -> do
+    fileFd <-
+      throwErrnoIfMinus1Retry "openFileFromDir" $
+      c_safe_openat dirFd f (ioModeToFlags iomode) 0o666
+    (fD, fd_type) <-
+      FD.mkFD
+        fileFd
+        iomode
+        Nothing {- no stat -}
+        False {- not a socket -}
+        False {- non_blocking --}
+         -- TODO: We want to investigate if we can do non-blocking here
+       `catchAny` \e -> do
```

Comment:
I'd use `onException` here instead, unless there's a specific reason you want to ignore async exceptions here.